File: packages/python/src/pico_store_lab/credentials.py
"""Account persistence through the platform credential store."""

from __future__ import annotations

import json
import sys

from keyring.backend import KeyringBackend

from pico_store_lab.protocol import PicoAuth

SERVICE = "dev.nkanf.picostore.python"
ACCOUNT = "current"


def _backend() -> KeyringBackend:
    # Select native stores directly: never load a configured plaintext/null backend.
    if sys.platform == "darwin":
        from keyring.backends.macOS import Keyring

        return Keyring()
    if sys.platform == "win32":
        from keyring.backends.Windows import WinVaultKeyring

        return WinVaultKeyring()
    if sys.platform.startswith("linux"):
        from keyring.backends.SecretService import Keyring

        return Keyring()
    raise RuntimeError("Sign-in requires Windows, macOS, or Linux.")


def save(auth: PicoAuth) -> None:
    """Save a usable account session."""
    if not auth.x_tt_token and not auth.cookies:
        raise ValueError("Sign-in returned no usable session. Please sign in again.")
    value = json.dumps({"uid": auth.uid, "x_tt_token": auth.x_tt_token, "cookies": auth.cookies})
    try:
        _backend().set_password(SERVICE, ACCOUNT, value)
    except Exception:
        raise RuntimeError(
            "Unable to save sign-in. Unlock the system credential store and try again."
        ) from None


def load() -> PicoAuth:
    """Read the current account session."""
    try:
        raw = _backend().get_password(SERVICE, ACCOUNT)
    except Exception:
        raise RuntimeError(
            "Unable to read sign-in. Unlock the system credential store and try again."
        ) from None
    if raw is None:
        raise ValueError("Please run pico-store-py login first.")
    try:
        data = json.loads(raw)
        if not isinstance(data, dict) or not isinstance(data.get("cookies", {}), dict):
            raise ValueError
        auth = PicoAuth(
            str(data.get("uid", "0")),
            str(data.get("x_tt_token", "")),
            {str(key): str(value) for key, value in data.get("cookies", {}).items()},
        )
        if not auth.x_tt_token and not auth.cookies:
            raise ValueError
    except (ValueError, TypeError):
        raise ValueError("Sign-in could not be read. Please sign in again.") from None
    return auth


def clear() -> None:
    """Remove the current account session."""
    try:
        backend = _backend()
        if backend.get_password(SERVICE, ACCOUNT) is not None:
            backend.delete_password(SERVICE, ACCOUNT)
    except Exception:
        raise RuntimeError(
            "Could not sign out. Unlock the system credential store and try again."
        ) from None

File: apps/desktop-rs/src/credentials.rs
use anyhow::{Result, anyhow, bail};
use keyring::{Entry, Error};
use pico_store_lab::PicoAuth;
use serde::{Deserialize, Serialize};

#[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "linux")))]
compile_error!("The desktop app requires macOS, Windows, or Linux");

#[derive(Clone, Serialize, Deserialize)]
pub struct Account {
    pub email: String,
    pub auth: PicoAuth,
}

fn entry() -> Result<Entry> {
    Entry::new("dev.nkanf.picostore.desktop", "current")
        .map_err(|_| anyhow!("Unable to open the system credential store."))
}

pub fn save(account: &Account) -> Result<()> {
    if account.auth.x_tt_token.is_empty() && account.auth.cookies.is_empty() {
        bail!("Sign-in returned no usable session.");
    }
    let value = serde_json::to_string(account)?;
    entry()?.set_password(&value).map_err(|_| {
        anyhow!("Unable to save sign-in. Unlock the system credential store and try again.")
    })
}

pub fn load() -> Result<Option<Account>> {
    let value = match entry()?.get_password() {
        Ok(value) => value,
        Err(Error::NoEntry) => return Ok(None),
        Err(_) => {
            bail!("Unable to read sign-in. Unlock the system credential store and try again.")
        }
    };
    let account: Account = serde_json::from_str(&value)
        .map_err(|_| anyhow!("Sign-in could not be read. Please sign in again."))?;
    if account.auth.x_tt_token.is_empty() && account.auth.cookies.is_empty() {
        bail!("Please sign in again.");
    }
    Ok(Some(account))
}

pub fn clear() -> Result<()> {
    match entry()?.delete_credential() {
        Ok(()) | Err(Error::NoEntry) => Ok(()),
        Err(_) => Err(anyhow!(
            "Could not sign out. Unlock the system credential store and try again."
        )),
    }
}
